Add generic EntityTeleportEvent

SizableShrimp opened the pull request:
Forge currently does not have a generic `EntityTeleportEvent` that can be used for catching any and all teleportation related to an `Entity`. This PR adds the functionality along with hooks into vanilla, enabling the events.

Currently, `TeleportCommand`, `SpreadPlayersCommand`, `EnderEntity`, `EnderPearl`, and `ChorusFruit` are all supported subclasses. Documentation has also been added to the generic event and all subclasses. `EnderTeleportEvent` has been deprecated for removal in 1.17 in favor of `EntityTeleportEvent.EnderEntity` and `EntityTeleportEvent.EnderPearl`.

A test mod is also provided that logs whenever an entity teleport event takes place. I also did local testing and confirmed that cancelling the event works in all supported subclasses.

I tried to keep my patch changes minimal, but please let me know if there is anything else I can do to minimize them further.

SizableShrimp committed:
Add generic EntityTeleportEvent and test mod

robotgryphon commented on patches/minecraft/net/minecraft/command/impl/TeleportCommand.java.patch:
```diff
@@ -1,5 +1,17 @@
+    }
+ 
+    private static void func_201127_a(CommandSource p_201127_0_, Entity p_201127_1_, ServerWorld p_201127_2_, double p_201127_3_, double p_201127_5_, double p_201127_7_, Set<SPlayerPositionLookPacket.Flags> p_201127_9_, float p_201127_10_, float p_201127_11_, @Nullable TeleportCommand.Facing p_201127_12_) throws CommandSyntaxException {
++      net.minecraft.util.math.vector.Vector3d target = net.minecraftforge.event.ForgeEventFactory.onEntityTeleportCommand(p_201127_1_, p_201127_3_, p_201127_5_, p_201127_7_);
++      if (target == null) return;
++      p_201127_3_ = target.field_72450_a;
++      p_201127_5_ = target.field_72448_b;
++      p_201127_7_ = target.field_72449_c;
```
Can this be traded for a cancelled event and a one-line version for the x-y-z values?
```java
if(!event) return;
p_201127_3_ = event.target.field_72450_a; p_201127_5_ = event.target.field_72448_b; p_201127_7_ = event.target.field_72449_c;
```

SizableShrimp replied:
I'll refactor this soon

SizableShrimp replied:
Fixed with 8b629ad.

robotgryphon commented on patches/minecraft/net/minecraft/command/impl/TeleportCommand.java.patch:
```diff
@@ -1,5 +1,17 @@
++      p_201127_3_ = target.field_72450_a;
++      p_201127_5_ = target.field_72448_b;
++      p_201127_7_ = target.field_72449_c;
+       BlockPos blockpos = new BlockPos(p_201127_3_, p_201127_5_, p_201127_7_);
+       if (!World.func_234935_k_(blockpos)) {
+          throw field_241077_a_.create();
```
Checking the position is loaded? Explain why this is added to the patch.

SizableShrimp replied:
I didn't add this, so I'm not sure how it appeared.

SizableShrimp replied:
Actually, this is automatically generated to show the relevant lines around the patch. This is automatic. Notice how there is no `+` or `-` at the start of the line, meaning this was not changed and is just being displayed.

TheCurle replied:
Ah yes, a patch of a patch. Truly the bane of our collective existence.

SizableShrimp committed:
Update factory methods and add documentation note

robotgryphon commented on patches/minecraft/net/minecraft/command/impl/SpreadPlayersCommand.java.patch:
```diff
@@ -0,0 +1,12 @@
+ 
+-         entity.func_223102_j((double)MathHelper.func_76128_c(spreadplayerscommand$position.field_198713_a) + 0.5D, (double)spreadplayerscommand$position.func_198710_a(p_241072_1_, p_241072_3_), (double)MathHelper.func_76128_c(spreadplayerscommand$position.field_198714_b) + 0.5D);
++         net.minecraftforge.event.entity.living.EntityTeleportEvent.SpreadPlayersCommand event = net.minecraftforge.event.ForgeEventFactory.onEntityTeleportSpreadPlayersCommand(entity, (double)MathHelper.func_76128_c(spreadplayerscommand$position.field_198713_a) + 0.5D, (double)spreadplayerscommand$position.func_198710_a(p_241072_1_, p_241072_3_), (double)MathHelper.func_76128_c(spreadplayerscom ...
++         if (!event.isCanceled()) entity.func_223102_j(event.getTargetX(), event.getTargetY(), event.getTargetZ());
```
Else condition for the event here? Shouldn't this perform an early exit or continue?

SizableShrimp replied:
The patch does not fully show the scope of this method execution. The code in question is actually executed in a loop for all entities selected by the selector or w/e in `/spreadplayers`. Exiting early would cause all entities after the one that is canceled from teleportation to also be canceled, which would be an unintended side effect. This way, entities can be selectively blocked from teleporting instead of all of them.

TheCurle commented on patches/minecraft/net/minecraft/entity/item/EnderPearlEntity.java.patch:
```diff
@@ -1,23 +1,26 @@
                 }
  
 -               entity.func_70634_a(this.func_226277_ct_(), this.func_226278_cu_(), this.func_226281_cx_());
-+               entity.func_70634_a(event.getTargetX(), event.getTargetY(), event.getTargetZ());
```
Is there any reason you can't use the event's positions here, just so the patch is smaller? And when the Ender event is scrapped in 1.17, it should be obvious to use the new "evt"'s position..

TheCurle commented on src/test/java/net/minecraftforge/debug/entity/EntityTeleportEventTest.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    @SubscribeEvent
+    public void registerCommands(EntityTeleportEvent event)
```
Bizarre name for a teleport listener..

TheCurle reviewed (approved):
Seems good. Some small optimisations are possible, but nothing world shattering.

SizableShrimp committed:
Fix event subscriber name

SizableShrimp committed:
Refactor ender teleport event post and reorder ender pearl teleport parameters

It should now be easier to remove `EnderTeleportEvent` in 1.17 by deleting the relevant lines in `ForgeEventFactory`.
`EntityTeleportEvent.EnderPearl` parameters were reordered to more obviously show `ServerPlayerEntity` as the entity designated to be teleported.

SizableShrimp commented:
Updated to reduce the patch size. `EnderPearlEntity`, `EndermanEntity`, and `ShulkerEntity` have all been refactored to instead use a similar call to the original `EnderTeleportEvent` to reduce changes. `ForgeEventFactory` has also been modified to post an `EnderTeleportEvent` first for the 2 related events (`EnderEntity` & `EnderPearl`), and then modify the event with the values returned from the post of this event on the event bus. Removing this in 1.17 should now be simpler needing only to have the event class itself deleted, and the relevant method + 2 lines in `ForgeEventFactory` removed as well. Comments have been added detailing which lines need to be removed in 1.17.

TheCurle reviewed (approved):
That's now a lot easier to read. The code seems good to me, and some testing shows it works.
